openssl-dynamic/src/main/c/jnilib.c

```diff
@@ -225,3 +152,271 @@ jint tcn_get_java_env(JNIEnv **env)
+     * APR version.
+     */
+    apr_version(&apv);
+    apvn = apv.major * 1000 + apv.minor * 100 + apv.patch;
+    if (apvn < 1201) {
+        tcn_Throw(env, "Unsupported APR version (%s)",
+                  apr_version_string());
+        return JNI_ERR;
+    }
+
+
+    /* Initialize global java.lang.String class */
+    TCN_LOAD_CLASS(env, jString_class, "java/lang/String", JNI_ERR);
+
+    TCN_GET_METHOD(env, jString_class, jString_init,
+                   "<init>", "([B)V", JNI_ERR);
+    TCN_GET_METHOD(env, jString_class, jString_getBytes,
+                   "getBytes", "()[B", JNI_ERR);
+
+    TCN_LOAD_CLASS(env, byteArrayClass, "[B", JNI_ERR);
+
+    char* keyMaterialClassName = netty_internal_tcnative_util_prepend(packagePrefix, "io/netty/internal/tcnative/CertificateRequestedCallback$KeyMaterial");
+    jclass keyMaterialClassLocal = (*env)->FindClass(env, keyMaterialClassName);
+    free(keyMaterialClassName);
+    keyMaterialClassName = NULL;
+    if (keyMaterialClassLocal == NULL) {
+        return JNI_ERR;
+    }
+    keyMaterialClass = (*env)->NewGlobalRef(env, keyMaterialClassLocal);        \
+
+    TCN_GET_FIELD(env, keyMaterialClass, keyMaterialCertificateChainFieldId,
+                   "certificateChain", "J", JNI_ERR);
+    TCN_GET_FIELD(env, keyMaterialClass, keyMaterialPrivateKeyFieldId,
+                   "privateKey", "J", JNI_ERR);
+
+
+    return TCN_JNI_VERSION;
+}
+
+void netty_internal_tcnative_Library_JNI_OnUnLoad(JNIEnv* env) {
+    if (tcn_global_pool) {
+        TCN_UNLOAD_CLASS(env, jString_class);
+        apr_terminate();
+    }
+
+    TCN_UNLOAD_CLASS(env, byteArrayClass);
+    TCN_UNLOAD_CLASS(env, keyMaterialClass);
+
+    netty_internal_tcnative_Buffer_JNI_OnUnLoad(env);
+    netty_internal_tcnative_NativeStaticallyReferencedJniMethods_JNI_OnUnLoad(env);
+    netty_internal_tcnative_SSL_JNI_OnUnLoad(env);
+    netty_internal_tcnative_SSLContext_JNI_OnUnLoad(env);
+}
+
+
+jint JNI_OnLoad(JavaVM* vm, void* reserved) {
+    JNIEnv* env;
+    if ((*vm)->GetEnv(vm, (void**) &env, TCN_JNI_VERSION) != JNI_OK) {
+        return JNI_ERR;
+    }
+
+    jint status = 0;
+    const char* name = NULL;
+#ifndef WIN32
+    Dl_info dlinfo;
+    // We need to use an address of a function that is uniquely part of this library, so choose a static
+    // function. See https://github.com/netty/netty/issues/4840.
+    if (!dladdr((void*) parsePackagePrefix, &dlinfo)) {
+        fprintf(stderr, "FATAL: netty-tcnative JNI call to dladdr failed!\n");
+        return JNI_ERR;
+    }
+    name = dlinfo.dli_fname;
+#else
+    HMODULE module = NULL;
+    if (GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, (void*) parsePackagePrefix, &module) == 0){
+        fprintf(stderr, "FATAL: netty-tcnative JNI call to GetModuleHandleExA failed!\n");
+        return JNI_ERR;
+    }
+
+    char dllPath[2048];
+    if (GetModuleFileNameA(module, dllPath, 2048) == 0) {
```

Comment:
https://msdn.microsoft.com/en-us/library/windows/desktop/ms683197(v=vs.85).aspx

> Windows XP:  The string is truncated to nSize characters and is not null-terminated.

I guess we should force the last index to NULL otherwise infinite loop ... or we could make sure the string utility methods don't rely upon NULL (e.g. `strlen`)